lib/swtpmtest.pm

```diff
@@ -0,0 +1,124 @@
+        version     => '2.0',
+        ssh_script  => 'ssh_script_2',
+        result_file => '/tmp/swtpm_2',
+    },
+};
+
+my $sample_cfg = {
+    uefi   => {vm_name => 'SLES15SP3-UEFI-swtpm',   sample_file => 'swtpm_uefi.xml'},
+    legacy => {vm_name => 'SLES15SP3-LEGACY-swtpm', sample_file => 'swtpm_legacy.xml'},
+};
+
+sub start_swtpm_vm {
+    my ($swtpm_ver, $swtpm_vm_type) = @_;
+    die "invalid swtpm or vm type parameter defined" unless ($guestvm_cfg->{$swtpm_ver} && $sample_cfg->{$swtpm_vm_type});
+
+    # Copy the sample configuration file and modify it based on the test requirement
+    my $guest_swtpm_ver = $guestvm_cfg->{$swtpm_ver};
+    my $guest_mode      = $sample_cfg->{$swtpm_vm_type};
+    my $vm_name         = $guest_mode->{vm_name};
+    my $smaple_xml      = $guest_mode->{sample_file};
+    my $guest_xml       = $guest_swtpm_ver->{xml_file};
+    my $swtpm_type      = $guest_swtpm_ver->{version};
+    assert_script_run("cd $image_path");
+    assert_script_run("cp $smaple_xml $guest_xml->{$swtpm_vm_type}");
+    assert_script_run(
+"sed -i \"/<\\/devices>/i\\    <tpm model='tpm-tis'>\\n      <backend type='emulator' version='$swtpm_type'\\/>\\n    <\\/tpm>\" $guest_xml->{$swtpm_vm_type}"
+    );
+
+    # Define the guest vm and start it
+    assert_script_run("virsh define $guest_xml->{$swtpm_vm_type}");
+    assert_script_run("virsh start $vm_name");
+}
+
+sub stop_swtpm_vm {
+    my $para = shift;
+    die "invalid vm type parameter defined" unless ($sample_cfg->{$para});
+
+    # For UEFI vm guest, we should add --nvram option
+    my $guest_mode = $sample_cfg->{$para};
+    my $vm_name    = $guest_mode->{vm_name};
+    assert_script_run("virsh destroy $vm_name");
+    if ($para eq "uefi") {
+        assert_script_run("virsh undefine $vm_name --nvram");
+    }
+    else {
+        assert_script_run("virsh undefine $vm_name");
+    }
+}
+
+sub swtpm_verify {
+    my $para = shift;
+    die "invalid swtpm parameter defined" unless ($guestvm_cfg->{$para});
+
+    # Login to the vm and run the commands to check tpm device
+    my $user   = 'root';
+    my $passwd = 'nots3cr3t';
+    my $ip_addr;
+    my $results = script_run("which arp");
+    if ($results) {
+        record_soft_failure("bsc#1180446 - can not install package 'net-tools-deprecated'");
+        $ip_addr = script_output("ip n | grep 192.168 | awk '{print \$1}'");
+    }
+    else {
+        $ip_addr = script_output("arp | grep 192.168 | awk '{print \$1}'");
+    }
+    my $guest_swtpm_ver = $guestvm_cfg->{$para};
+    my $result_file     = $guest_swtpm_ver->{result_file};
+    my $ssh_script      = "$image_path/$guest_swtpm_ver->{ssh_script}";
+    assert_script_run("$ssh_script $ip_addr $user $passwd > $result_file");
+    assert_script_run("cat $result_file | grep tpm0");
+    if ($para eq "swtpm_1") {
+        assert_script_run("cat $result_file | grep 'TPM 1.2 Version'");
```

Comment:
Just a minor nitpick, but `grep 'TPM 1.2 Version' $result_file` is simpler and achieves the same.

Reply:
Thanks, I can't agree with you anymore. let me fix it

Reply:
fixed